src/wallets/router/routes.component.tsx

```diff
@@ -0,0 +1,77 @@
+import React, { useEffect, useMemo, type PropsWithChildren } from "react";
+import { Switch, Route as Woute } from "wouter";
+import { Page } from "~components/page/page.component";
+import type {
+  CommonRouteProps,
+  RouteConfig
+} from "~wallets/router/router.types";
+import { useLocation } from "~wallets/router/router.utils";
+
+export interface RoutesProps {
+  routes: RouteConfig[];
+  diffLocation?: boolean;
+  pageComponent?: React.ComponentType<PropsWithChildren>;
+}
+
+// TODO: Consider adding `<AnimatePresence>` and `<BodyScroller>` inside here.
+
+// TODO: Consider adding a prop to `RouteConfig.parseParams` to parse
+// params globally inside `PageWithComponent` (e.g. to replace the `Number()`)
+// conversions in the Welcome views.
+
+export function Routes({
+  routes,
+  diffLocation = false,
+  pageComponent
+}: RoutesProps) {
+  const { location } = useLocation();
+
+  // In development, check there are no duplicate routes (paths):
+
+  if (process.env.NODE_ENV === "development") {
+    useEffect(() => {
+      const uniqueRoutes = new Set();
+
+      routes.forEach(({ path }) => {
+        if (uniqueRoutes.has(path))
+          throw new Error(`Duplicate route "${path}"`);
+
+        uniqueRoutes.add(path);
+      });
+    }, [routes]);
+  }
+
+  const memoizedRoutes = useMemo(() => {
+    return (
+      <Switch>
+        {routes.map((route) => {
+          const Component = route.component;
+          const PageComponent =
+            pageComponent === null ? React.Fragment : pageComponent || Page;
+
+          // TODO: Async-loaded components?
+
+          const PageWithComponent: React.ComponentType<CommonRouteProps> = (
+            props
+          ) => {
+            return (
+              <PageComponent>
+                <Component {...props} />
+              </PageComponent>
+            );
+          };
+
+          return (
+            <Woute
+              key={route.key || route.path}
+              path={route.path}
+              component={PageWithComponent}
+            />
+          );
+        })}
+      </Switch>
+    );
+  }, [routes, diffLocation ? location : undefined]);
```

Comment:
Diffing the location makes different views that use the same component to show a transition between the two (e.g. selecting an `AuthRequest` of the same type in the Auth popup).